compiler/AST/ModuleSymbol.cpp

```diff
@@ -348,8 +348,19 @@ void ModuleSymbol::printDocs(std::ostream* file,
       return;
     }
     *file << ".. module:: " << this->docsName() << std::endl;
-
-    if (this->doc != NULL) {
+    // Don't index internal modules since that will make them show up
+    // in the module index (chpl-modindex.html).  This has the side
+    // effect of making references to the :mod: tag for the module
+    // illegal, which is appropriate since the modules are not
+    // user-facing.
+    if ((strncmp(this->name, "Chapel", 6) == 0 && strcmp(this->name, "ChapelIO") != 0) ||
+        strcmp(this->name, "Atomics") == 0 ||
+        strcmp(this->name, "Bytes") == 0 ||
+        strcmp(this->name, "OwnedObject") == 0 ||
+        strcmp(this->name, "SharedObject") == 0 ||
+        strcmp(this->name, "String") == 0) {
+      *file << "   :noindex:" << std::endl;
+    } else if (this->doc != NULL) {
```

Comment:
@lydia-duncan : This is the code that relates to my "Can chpldoc know that a module is internal" question the other day.  Would you give it a review and let me know whether you're OK with it going in?  (it's a bit fragile, but the only real consequence of getting the logic wrong is that a module will either appear correctly or incorrectly in our module index).

Reply:
I think we could do this.  I'd also be tempted to make a specific pragma and apply that, since that will make it less likely we'll need to adjust the compiler again if we miss a case or change our minds or something.

Reply:
I just pushed an improvement to this that looks at the module's filename and squashes it when it starts with "internal/", which feels better though still not ideal.  In the commit message I pointed out that the main downside of this is that it will similarly squash modules passed to `chpldoc` by end users that happen to have a filename of `internal/whatever.chpl`, but it's actually not quite as bad as this—they also have to use `--no-html`.

One other way I could imagine tightening this up would be to pass the full path to the internal modules via the Makefile and to have `chpldoc` see that the modules are in `$CHPL_HOME/modules/internal/`.  I just can't recall how well or poorly the compiler can reason about CHPL_HOME offhand...

I agree that introducing a pragma would be more principled but am a little hesitant to do so given that we haven't needed such a pragma up until now, the number of files it'll touch, and the need to rely on future internal modules to similarly add the pragma.

Reply:
Sounds good!

Reply:
OK, I think the commit I just pushed should do this in an airtight way:  The Makefile passes in the absolute paths to the internal modules we want to document, and the docs.cpp pass looks for files in the same directory and tags them as MOD_INTERNAL.

Reply:
Even better :)